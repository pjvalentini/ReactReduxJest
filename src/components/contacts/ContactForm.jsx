import React from "react";
import PropTypes from "prop-types";
import TextInput from "../common/TextInput.jsx";
import SelectInput from "../common/SelectInput.jsx";

const ContactForm = ({
  contact,
  creators,
  onSave,
  onChange,
  saving = false,
  errors = {}
}) => {
  return (
    <form onSubmit={onSave}>
      <h2>{contact.id ? "Edit" : "Add"} Contact</h2>
      {errors.onSave && (
        <div className="alert alert-danger" role="alert">
          {errors.onSave}
        </div>
      )}
      <TextInput
        name="name"
        label="Name"
        value={contact.name}
        onChange={onChange}
        error={errors.name}
      />

      <TextInput
        name="category"
        label="Email"
        value={contact.category}
        onChange={onChange}
        error={errors.category}
      />

      <SelectInput
        name="creatorId"
        label="Created By"
        value={contact.creatorId || ""}
        defaultOption="Select Creator"
        options={creators.map(creator => ({
          value: creator.id,
          text: creator.name
        }))}
        onChange={onChange}
        error={errors.creator}
      />

      <button type="submit" disabled={saving} className="btn btn-primary">
        {saving ? "Saving..." : "Save"}
      </button>
    </form>
  );
};

ContactForm.propTypes = {
  creators: PropTypes.array.isRequired,
  contact: PropTypes.object.isRequired,
  errors: PropTypes.object,
  onSave: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
  saving: PropTypes.bool
};

export default ContactForm;
