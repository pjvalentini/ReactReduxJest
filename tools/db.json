{
  "contacts": [
    {
      "id": 1,
      "name": "John Doe",
      "slug": "John Doe",
      "creatorId": 7,
      "category": "John@Doe.com"
    },
    {
      "id": 2,
      "name": "Jane Doe",
      "slug": "Jane Doe",
      "creatorId": 1,
      "category": "Jane@Doe.com"
    },
    {
      "id": 3,
      "name": "Bill Clinton",
      "slug": "Bill Clinton",
      "creatorId": 2,
      "category": "potus@gmail.com"
    },
    {
      "id": 4,
      "name": "LeBron James",
      "slug": "Lebron James",
      "creatorId": 2,
      "category": "kingJames@gmail.com"
    },
    {
      "id": 5,
      "name": "Wayne Gretzky",
      "slug": "Wayne Gretzky",
      "creatorId": 1,
      "category": "greatone@gmail.com"
    },
    {
      "id": 6,
      "name": "Ted Dead",
      "slug": "Ted Dead",
      "creatorId": 1,
      "category": "Ted@Dead.com"
    },
    {
      "id": 7,
      "name": "Wilt Chamberlin",
      "slug": "Wilt Chamberlin",
      "creatorId": 1,
      "category": "wilt@theStilt"
    },
    {
      "id": 8,
      "name": "Sam Ashe",
      "slug": "Sam Ashe",
      "creatorId": 1,
      "category": "Sam@Ashe.com"
    },
    {
      "id": 9,
      "name": "Mario Mario",
      "slug": "Mario Mario",
      "creatorId": 3,
      "category": "marioM@gmail.com"
    },
    {
      "id": 10,
      "name": "Scottie Pippen",
      "slug": "Scottie Pippen",
      "creatorId": 1,
      "category": "SP@outlook.com"
    }
  ],
  "creators": [
    {
      "id": 1,
      "name": "Lea Collins"
    },
    {
      "id": 2,
      "name": "Kent Elmer"
    },
    {
      "id": 3,
      "name": "PJ Valentini"
    },
    {
      "id": 4,
      "name": "Paula Speakman"
    },
    {
      "id": 5,
      "name": "Robert Blackstone"
    },
    {
      "id": 6,
      "name": "Adam Cornett"
    },
    {
      "id": 7,
      "name": "Dan Jewell"
    },
    {
      "id": 8,
      "name": "Jeff Riedinger"
    },
    {
      "id": 9,
      "name": "Diva Payne"
    },
    {
      "id": 10,
      "name": "Amy Johnson"
    }
  ]
}