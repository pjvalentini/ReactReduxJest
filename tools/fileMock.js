// Mocks file imports for Jest. Suggested by https://jestjs.io/docs/en/webpack
module.export = "test-file-stub";
