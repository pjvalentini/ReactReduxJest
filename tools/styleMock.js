// This file mocks css imports for Jest. Suggested by https://jestjs.io/docs/en/webpack
module.exports = {};
