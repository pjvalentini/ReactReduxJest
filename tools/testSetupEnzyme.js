import { configure } from "enzyme";
// configures Adapter to work with react 16
import Adapter from "enzyme-adapter-react-16";
configure({ adapter: new Adapter() });
