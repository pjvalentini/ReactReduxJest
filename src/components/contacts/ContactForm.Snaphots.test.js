import React from "react";
import ContactForm from "./ContactForm.jsx";
import renderer from "react-test-renderer";
import { contacts, creators } from "../../../tools/mockData";

// Test 1: Assure that the label on the save button is properly set when we set the save prop to true.
it("sets save button label to 'Saving...' when saving is true", () => {
  const tree = renderer.create(
    <ContactForm
      contact={contacts[0]}
      creators={creators}
      onSave={jest.fn()}
      onChange={jest.fn()}
      saving
    />
  );
  expect(tree).toMatchSnapshot();
});

// Test 2: Assure that the label on the save button is properly set when we set the save prop to false.
it("sets save button label to 'Save' when saving is false", () => {
  const tree = renderer.create(
    <ContactForm
      contact={contacts[0]}
      creators={creators}
      onSave={jest.fn()}
      onChange={jest.fn()}
      saving={false}
    />
  );
  expect(tree).toMatchSnapshot();
});
