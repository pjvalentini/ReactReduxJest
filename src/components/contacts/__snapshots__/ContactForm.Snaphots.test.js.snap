// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sets save button label to 'Save' when saving is false 1`] = `
<form
  onSubmit={[MockFunction]}
>
  <h2>
    Edit
     Contact
  </h2>
  <div
    className="form-group"
  >
    <label
      htmlFor="name"
    >
      Name
    </label>
    <div
      className="field"
    >
      <input
        className="form-control"
        name="name"
        onChange={[MockFunction]}
        type="text"
        value="John Doe"
      />
    </div>
  </div>
  <div
    className="form-group"
  >
    <label
      htmlFor="category"
    >
      Email
    </label>
    <div
      className="field"
    >
      <input
        className="form-control"
        name="category"
        onChange={[MockFunction]}
        type="text"
        value="John@Doe.com"
      />
    </div>
  </div>
  <div
    className="form-group"
  >
    <label
      htmlFor="creatorId"
    >
      Created By
    </label>
    <div
      className="field"
    >
      <select
        className="form-control"
        name="creatorId"
        onChange={[MockFunction]}
        value={3}
      >
        <option
          value=""
        >
          Select Creator
        </option>
        <option
          value={1}
        >
          Lea Collins
        </option>
        <option
          value={2}
        >
          Kent Elmer
        </option>
        <option
          value={3}
        >
          PJ Valentini
        </option>
        <option
          value={4}
        >
          Paula Speakman
        </option>
        <option
          value={5}
        >
          Robert Blackstone
        </option>
        <option
          value={6}
        >
          Adam Cornett
        </option>
        <option
          value={7}
        >
          Dan Jewell
        </option>
        <option
          value={8}
        >
          Jeff Riedinger
        </option>
        <option
          value={9}
        >
          Diva Payne
        </option>
        <option
          value={10}
        >
          Amy Johnson
        </option>
      </select>
    </div>
  </div>
  <button
    className="btn btn-primary"
    disabled={false}
    type="submit"
  >
    Save
  </button>
</form>
`;

exports[`sets save button label to 'Saving...' when saving is true 1`] = `
<form
  onSubmit={[MockFunction]}
>
  <h2>
    Edit
     Contact
  </h2>
  <div
    className="form-group"
  >
    <label
      htmlFor="name"
    >
      Name
    </label>
    <div
      className="field"
    >
      <input
        className="form-control"
        name="name"
        onChange={[MockFunction]}
        type="text"
        value="John Doe"
      />
    </div>
  </div>
  <div
    className="form-group"
  >
    <label
      htmlFor="category"
    >
      Email
    </label>
    <div
      className="field"
    >
      <input
        className="form-control"
        name="category"
        onChange={[MockFunction]}
        type="text"
        value="John@Doe.com"
      />
    </div>
  </div>
  <div
    className="form-group"
  >
    <label
      htmlFor="creatorId"
    >
      Created By
    </label>
    <div
      className="field"
    >
      <select
        className="form-control"
        name="creatorId"
        onChange={[MockFunction]}
        value={3}
      >
        <option
          value=""
        >
          Select Creator
        </option>
        <option
          value={1}
        >
          Lea Collins
        </option>
        <option
          value={2}
        >
          Kent Elmer
        </option>
        <option
          value={3}
        >
          PJ Valentini
        </option>
        <option
          value={4}
        >
          Paula Speakman
        </option>
        <option
          value={5}
        >
          Robert Blackstone
        </option>
        <option
          value={6}
        >
          Adam Cornett
        </option>
        <option
          value={7}
        >
          Dan Jewell
        </option>
        <option
          value={8}
        >
          Jeff Riedinger
        </option>
        <option
          value={9}
        >
          Diva Payne
        </option>
        <option
          value={10}
        >
          Amy Johnson
        </option>
      </select>
    </div>
  </div>
  <button
    className="btn btn-primary"
    disabled={true}
    type="submit"
  >
    Saving...
  </button>
</form>
`;
